### PR TITLE
MAGETWO-2745: added appEmulation Plugin for the $sitemap->generateXml() for generating the sitemap via cron and backend

### LINKED ISSUE #29605: [Issue] MAGETWO-2745: added appEmulation Plugin for the $sitemap->generateXml() for generating the sitemap via cron and backend
This issue is automatically created based on existing pull request: magento/magento2#27476: MAGETWO-2745: added appEmulation Plugin for the $sitemap->generateXml() for generating the sitemap via cron and backend

---------

### Description (*)
I found almost the same issue here: #24605, here you generate the sitemap via backend.
In my case i let generate the sitemap with cron schedule. The problem is i get a wrong cache path, because the \Magento\Sitemap\Model\Observer:scheduledGenerateSitemaps() does not implement the appEmulation->stopEnvironmentEmulation(); like it was fixed in the issue #24605

### Manual testing scenarios (*)

1. Just generate a sitemap via backend
2. Generate a sitemap via cron 
3. You will see, there are two different cache paths 

for example via backend: 
```
/media/catalog/product/cache/cb8ca2a9db68e4f593ce69a9ee07e2b0/p/r/test.png
```

for example via cron:
```
/media/catalog/product/cache/9b58a33b64c72152de346a8490e93157/p/r/test.png
```

with the appEmulation inside the observer, cron and backend generates the same cache path.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by dimaMackert**
### Description (*)
I found almost the same issue here: #24605, here you generate the sitemap via backend.
In my case i let generate the sitemap with cron schedule. The problem is i get a wrong cache path, because the \Magento\Sitemap\Model\Observer:scheduledGenerateSitemaps() does not implement the appEmulation->stopEnvironmentEmulation(); like it was fixed in the issue #24605

### Manual testing scenarios (*)

1. Just generate a sitemap via backend
2. Generate a sitemap via cron 
3. You will see, there are two different cache paths 

for example via backend: 
```
/media/catalog/product/cache/cb8ca2a9db68e4f593ce69a9ee07e2b0/p/r/test.png
```

for example via cron:
```
/media/catalog/product/cache/9b58a33b64c72152de346a8490e93157/p/r/test.png
```

with the appEmulation inside the observer, cron and backend generates the same cache path.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#29605: MAGETWO-2745: added appEmulation Plugin for the $sitemap->generateXml() for generating the sitemap via cron and backend

### CODE REVIEW

File: app/code/Magento/Sitemap/Plugin/StoreEmulation.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+   public function __construct(Emulation $appEmulation)
+   {
+       $this->appEmulation = $appEmulation;
+   }
+
+   /**
+    * @param Sitemap $subject
+    * @see \Magento\Sitemap\Model\Sitemap::generateXml
+    */
+   public function beforeGenerateXml(Sitemap $subject)
```
Comment (kandy) [inline app/code/Magento/Sitemap/Plugin/StoreEmulation.php]: Better to use around plugin here
and write code in the way like 
```
 $this->appEmulation->startEnvironmentEmulation(
                    $sitemap->getStoreId(),
                    Area::AREA_FRONTEND,
                    true
                );
             try {
                retrun $proceed();
             } finally {
                $this->appEmulation->stopEnvironmentEmulation();
             }
```

Comment (kandy) [reply app/code/Magento/Sitemap/Plugin/StoreEmulation.php]: And actually why do not put this directly in generateXml method

File: app/code/Magento/Sitemap/Model/Observer.php [hunk 2]
```diff
@@ -105,7 +117,13 @@ public function scheduledGenerateSitemaps()
...
+                $this->appEmulation->startEnvironmentEmulation(
+                    $sitemap->getStoreId(),
+                    Area::AREA_FRONTEND,
+                    true
+                );
                 $sitemap->generateXml();
+                $this->appEmulation->stopEnvironmentEmulation();
             } catch (\Exception $e) {
                 $errors[] = $e->getMessage();
             }
```
Comment (kandy) [inline app/code/Magento/Sitemap/Model/Observer.php:129]:
```suggestion
            } catch (\Exception $e) {
                $errors[] = $e->getMessage();
            } finally {
                $this->appEmulation->stopEnvironmentEmulation();
            }
```

File: app/code/Magento/Sitemap/Plugin/StoreEmulation.php (hunk 1 above)
Comment (Stepa4man) [reply app/code/Magento/Sitemap/Plugin/StoreEmulation.php]: > And actually why do not put this directly in generateXml method

@kandy why do you think Sitemap itself should be responsible for modifying the Application state? I'd say it should be done by the observer.

File: app/code/Magento/Sitemap/Model/Observer.php [hunk 3]
```diff
@@ -105,7 +117,13 @@ public function scheduledGenerateSitemaps()
         foreach ($collection as $sitemap) {
             /* @var $sitemap \Magento\Sitemap\Model\Sitemap */
             try {
+                $this->appEmulation->startEnvironmentEmulation(
```
Comment (Stepa4man) [inline app/code/Magento/Sitemap/Model/Observer.php:120]: I would suggest moving area emulation outside of the foreach statement

**Review by Stepa4man — CHANGES_REQUESTED** [review]
Hi @dimaMackert 
Thanks for your contribution

Please sign CLA and consider suggested changes.

**Comment by Stepa4man** [comment]
@VladimirZaets, could you please check what's going on with tests right now?
I can see that fail functional tests don't relate to PR itself

**Comment by sidolov** [comment]
@magento create issue

**Comment by sidolov** [comment]
@dimaMackert , I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

**Pull request closed without merging**